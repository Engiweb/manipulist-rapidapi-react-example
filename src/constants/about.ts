export const ENGIWEB_WEBSITE = "https://engiweb.com";
export const MANIPULIST_DOCS = "https://manipulist.com/docs";
