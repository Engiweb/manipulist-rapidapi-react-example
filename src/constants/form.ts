export const DEFAULT_INPUT_TEXT = `The quick brown dog jumped over the 
lazy fox. The quick brown dog jumped 
over the lazy fox. The quick brown dog.
jumped over the lazy fox. The quick
brown dog jumped over the lazy fox.`;

export type TToolsParamObj = {
  type: string;
  description: string;
  enum?: string[];
};

export type TTToolObj = {
  required?: string[];
  param1?: TToolsParamObj;
  param2?: TToolsParamObj;
};

type TToolsForm = {
  [key: string]: TTToolObj;
};

export const TOOLS_FORM: TToolsForm = {
  "add-incremental-number": {
    required: ["param1"],
    param1: {
      type: "integer",
      description: "Starting number",
    },
    param2: {
      type: "string",
      description: "Incremental number position",
      enum: ["atStart", "atEnd"],
    },
  },
  "extract-rows": {
    required: ["param1"],
    param1: {
      description: "List or range of rows",
      type: "range",
    },
  },
  "find-and-replace": {
    required: ["param1"],
    param1: {
      description: "Replace",
      type: "string",
    },
    param2: {
      description: "With",
      type: "string",
    },
  },
  "remove-duplicate-lines": {},
  "sort-lines": {
    param1: {
      description: "Sorting method",
      type: "string",
      enum: ["az", "za", "reverse", "shuffle"],
    },
  },
};

export const TOOLS_LABELS = Object.keys(TOOLS_FORM);

export const LB_OPTIONS = ["lf", "crlf"];

export const INPUT_TYPE = ["File", "Text"];
