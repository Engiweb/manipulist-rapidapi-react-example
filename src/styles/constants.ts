export const COLORS = {
  background: "black",
  font: "white",
  primary: "#14e978",
  headerBg: "#222831",
};
