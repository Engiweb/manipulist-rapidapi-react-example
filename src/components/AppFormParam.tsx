import { FieldValues, UseFormRegister } from "react-hook-form";
import { TTToolObj } from "../constants/form";
import integersListRangeChecker from "../utils/integersListRangeChecker";
import styled from "styled-components";

const ErrorMessage = styled.span`
  color: red;
  margin-top: 5px;
  display: block;
`;

type TProps = {
  register: UseFormRegister<FieldValues>;
  tool: TTToolObj;
  errors: {
    [key: string]: string;
  };
  name: "param1" | "param2";
};

const AppFormParam = ({ register, tool, errors, name }: TProps) => {
  const param = tool[name];

  const required = tool.required?.includes(name);

  let field;

  switch (param?.type) {
    case "integer":
      field = (
        <input
          id={name}
          defaultValue={1}
          type="number"
          placeholder="Insert number"
          min={0}
          step={1}
          {...register(name, {
            required,
            min: 1,
            max: Number.MAX_SAFE_INTEGER,
          })}
        />
      );
      break;

    case "range":
      field = (
        <input
          id={name}
          placeholder="Insert range e.g. 1,3-5,7"
          defaultValue="1,2-3"
          {...register(name, {
            required,
            validate: (value) => integersListRangeChecker(value),
          })}
        />
      );
      break;

    default:
      if (param?.enum) {
        field = (
          <select
            placeholder="Choose from list"
            id={name}
            {...register(name, { required })}
          >
            {param?.enum?.map((tool) => (
              <option key={`${name}-${tool}`} value={tool}>
                {tool}
              </option>
            ))}
          </select>
        );
        break;
      }

      field = (
        <input
          id={name}
          placeholder="Insert text"
          defaultValue="text"
          {...register(name, {
            required,
            validate: (value) => (required ? value !== "" : true),
          })}
        />
      );
      break;
  }

  return (
    <>
      <label htmlFor="param">
        {param?.description} ({name})
      </label>
      {field}
      {errors[name] && (
        <ErrorMessage>
          {param?.type === "integer"
            ? "Insert a valid integer"
            : param?.type === "range"
            ? "Insert a valid range e.g. 1,3-5,7"
            : "This field is required"}
        </ErrorMessage>
      )}
    </>
  );
};

export default AppFormParam;
