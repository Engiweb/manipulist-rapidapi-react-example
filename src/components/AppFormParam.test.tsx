import { render, screen } from "@testing-library/react";
import { TOOLS_FORM } from "../constants/form";
import AppFormParam from "./AppFormParam";

const register = jest.fn();

describe("AppFormParam", () => {
  afterEach(() => jest.resetAllMocks());

  test("should work when add-incremental-number and param1", () => {
    const { asFragment } = render(
      <AppFormParam
        register={register}
        tool={TOOLS_FORM["add-incremental-number"]}
        errors={{}}
        name="param1"
      />
    );

    expect(asFragment).toMatchSnapshot();
  });

  test("should work when add-incremental-number and param2", () => {
    const { asFragment } = render(
      <AppFormParam
        register={register}
        tool={TOOLS_FORM["add-incremental-number"]}
        errors={{}}
        name="param2"
      />
    );

    expect(asFragment).toMatchSnapshot();
  });
});
