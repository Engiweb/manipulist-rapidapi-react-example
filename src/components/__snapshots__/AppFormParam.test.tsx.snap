// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppFormParam should work when add-incremental-number and param1 1`] = `[Function]`;

exports[`AppFormParam should work when add-incremental-number and param2 1`] = `[Function]`;
